File: lib/java/src/test/java/com/truveta/opentoken/tokens/tokenizer/PassthroughTokenizerTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens.tokenizer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.ArrayList;
import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import com.truveta.opentoken.tokens.Token;
import com.truveta.opentoken.tokentransformer.EncryptTokenTransformer;
import com.truveta.opentoken.tokentransformer.HashTokenTransformer;
import com.truveta.opentoken.tokentransformer.TokenTransformer;

class PassthroughTokenizerTest {
    private TokenTransformer hashTransformerMock;
    private TokenTransformer encryptTransformerMock;
    private PassthroughTokenizer tokenizer;

    @BeforeEach
    void setUp() {
        // Mocking TokenTransformer implementations (Hash and Encrypt)
        hashTransformerMock = Mockito.mock(HashTokenTransformer.class);
        encryptTransformerMock = Mockito.mock(EncryptTokenTransformer.class);

        // List of transformers to pass to PassthroughTokenizer
        List<TokenTransformer> transformers = new ArrayList<>();
        transformers.add(hashTransformerMock);
        transformers.add(encryptTransformerMock);

        // Instantiate the tokenizer with mocked transformers
        tokenizer = new PassthroughTokenizer(transformers);
    }

    @Test
    void testTokenize_NullOrEmptyInput_ReturnsEmptyString() throws Exception {
        String resultNull = tokenizer.tokenize(null); // Test for null input
        assertEquals(Token.BLANK, resultNull);

        String resultEmpty = tokenizer.tokenize(""); // Test for empty string input
        assertEquals(Token.BLANK, resultEmpty);

        String resultBlank = tokenizer.tokenize("   "); // Test for input with only whitespace
        assertEquals(Token.BLANK, resultBlank);
    }

    @Test
    void testTokenize_ValidInput_ReturnsUnchangedValue() throws Exception {
        String inputValue = "test-input";

        // Mock the transformations to simulate behavior of TokenTransformers
        when(hashTransformerMock.transform(anyString())).thenReturn(inputValue);
        when(encryptTransformerMock.transform(inputValue)).thenReturn("encrypted-token");

        String result = tokenizer.tokenize(inputValue); // Call the tokenize method

        // Verify the transformers were called with the original input value
        verify(hashTransformerMock).transform(inputValue);
        verify(encryptTransformerMock).transform(inputValue);

        assertEquals("encrypted-token", result); // Check the final result after applying the transformers
    }

    @Test
    void testTokenize_ValidInput_NoTransformers_ReturnsOriginalValue() throws Exception {
        String inputValue = "test-input";

        tokenizer = new PassthroughTokenizer(new ArrayList<>()); // Recreate tokenizer with no transformers

        String result = tokenizer.tokenize(inputValue); // Call the tokenize method

        assertEquals(inputValue, result); // Verify that the result is the original value unchanged
    }

    @Test
    void testTokenize_ValidInput_TransformerThrowsException() throws Exception {
        String inputValue = "test-input";

        // Mock the first transformer to throw an exception
        when(hashTransformerMock.transform(anyString())).thenThrow(new RuntimeException("Transform error"));

        // Call the tokenize method and assert it propagates the exception
        Exception exception = assertThrows(Exception.class, () -> {
            tokenizer.tokenize(inputValue);
        });

        assertEquals("Transform error", exception.getMessage());
    }

    @Test
    void testTokenize_MultipleTransformers_AppliesInOrder() throws Exception {
        String inputValue = "original-value";
        String afterFirstTransform = "after-first";
        String afterSecondTransform = "after-second";

        // Mock the transformers to apply sequential transformations
        when(hashTransformerMock.transform(inputValue)).thenReturn(afterFirstTransform);
        when(encryptTransformerMock.transform(afterFirstTransform)).thenReturn(afterSecondTransform);

        String result = tokenizer.tokenize(inputValue);

        // Verify transformers were called in order
        verify(hashTransformerMock).transform(inputValue);
        verify(encryptTransformerMock).transform(afterFirstTransform);

        assertEquals(afterSecondTransform, result);
    }

    @Test
    void testTokenize_SpecialCharacters_ReturnsUnchanged() throws Exception {
        String inputValue = "special!@#$%^&*()_+-=[]{}|;':\",./<>?";

        tokenizer = new PassthroughTokenizer(new ArrayList<>()); // No transformers

        String result = tokenizer.tokenize(inputValue);

        assertEquals(inputValue, result); // Special characters should pass through unchanged
    }

    @Test
    void testTokenize_UnicodeCharacters_ReturnsUnchanged() throws Exception {
        String inputValue = "Hello 世界 🌍";

        tokenizer = new PassthroughTokenizer(new ArrayList<>()); // No transformers

        String result = tokenizer.tokenize(inputValue);

        assertEquals(inputValue, result); // Unicode characters should pass through unchanged
    }
}

File: lib/python/opentoken/src/test/opentoken/tokens/tokenizer/passthrough_tokenizer_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import pytest
from unittest.mock import Mock
from opentoken.tokens.tokenizer.passthrough_tokenizer import PassthroughTokenizer


class TestPassthroughTokenizer:
    """Test cases for PassthroughTokenizer."""

    @pytest.fixture
    def setup(self):
        """Set up test fixtures."""
        # Mocking TokenTransformer implementations
        self.hash_transformer_mock = Mock()
        self.encrypt_transformer_mock = Mock()

        # List of transformers to pass to PassthroughTokenizer
        transformers = [
            self.hash_transformer_mock,
            self.encrypt_transformer_mock,
        ]

        # Instantiate the tokenizer with mocked transformers
        self.tokenizer = PassthroughTokenizer(transformers)

    def test_tokenize_null_or_empty_input_returns_empty_string(self, setup):
        """Test that None or empty input returns EMPTY constant."""
        result_none = self.tokenizer.tokenize(None)  # Test for None input
        assert result_none == PassthroughTokenizer.EMPTY

        result_empty = self.tokenizer.tokenize("")  # Test for empty string input
        assert result_empty == PassthroughTokenizer.EMPTY

        result_blank = self.tokenizer.tokenize("   ")  # Test for input with only whitespace
        assert result_blank == PassthroughTokenizer.EMPTY

    def test_tokenize_valid_input_returns_unchanged_value(self, setup):
        """Test that valid input is passed through with transformers applied."""
        input_value = "test-input"

        # Mock the transformations to simulate behavior of TokenTransformers
        self.hash_transformer_mock.transform.return_value = input_value
        self.encrypt_transformer_mock.transform.return_value = "encrypted-token"

        result = self.tokenizer.tokenize(input_value)  # Call the tokenize method

        # Verify the transformers were called with the original input value
        self.hash_transformer_mock.transform.assert_called_once_with(input_value)
        self.encrypt_transformer_mock.transform.assert_called_once_with(input_value)

        assert result == "encrypted-token"  # Check the final result after applying transformers

    def test_tokenize_valid_input_no_transformers_returns_original_value(self):
        """Test that without transformers, the original value is returned unchanged."""
        input_value = "test-input"

        tokenizer = PassthroughTokenizer([])  # Recreate tokenizer with no transformers

        result = tokenizer.tokenize(input_value)  # Call the tokenize method

        assert result == input_value  # Verify that the result is the original value unchanged

    def test_tokenize_valid_input_transformer_throws_exception(self, setup):
        """Test that transformer exceptions are propagated."""
        input_value = "test-input"

        # Mock the first transformer to throw an exception
        self.hash_transformer_mock.transform.side_effect = RuntimeError("Transform error")

        # Call the tokenize method and assert it propagates the exception
        with pytest.raises(RuntimeError) as exc_info:
            self.tokenizer.tokenize(input_value)

        assert str(exc_info.value) == "Transform error"

    def test_tokenize_multiple_transformers_applies_in_order(self, setup):
        """Test that multiple transformers are applied in sequence."""
        input_value = "original-value"
        after_first_transform = "after-first"
        after_second_transform = "after-second"

        # Mock the transformers to apply sequential transformations
        self.hash_transformer_mock.transform.return_value = after_first_transform
        self.encrypt_transformer_mock.transform.return_value = after_second_transform

        result = self.tokenizer.tokenize(input_value)

        # Verify transformers were called in order
        self.hash_transformer_mock.transform.assert_called_once_with(input_value)
        self.encrypt_transformer_mock.transform.assert_called_once_with(after_first_transform)

        assert result == after_second_transform

    def test_tokenize_special_characters_returns_unchanged(self):
        """Test that special characters pass through unchanged."""
        input_value = "special!@#$%^&*()_+-=[]{}|;':\",./<>?"

        tokenizer = PassthroughTokenizer([])  # No transformers

        result = tokenizer.tokenize(input_value)

        assert result == input_value  # Special characters should pass through unchanged

    def test_tokenize_unicode_characters_returns_unchanged(self):
        """Test that Unicode characters pass through unchanged."""
        input_value = "Hello 世界 🌍"

        tokenizer = PassthroughTokenizer([])  # No transformers

        result = tokenizer.tokenize(input_value)

        assert result == input_value  # Unicode characters should pass through unchanged
